internal/inventory/tag_test.go

```diff
@@ -0,0 +1,196 @@
+}
+
+func TestNewTag(t *testing.T) {
+	type args struct {
+		epc string
+	}
+	tests := []struct {
+		name string
+		args args
+		want *Tag
+	}{
+		{
+			name: "OK",
+			args: args{epc: "test"},
+			want: tagHelper("test", TagState(Unknown)),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got := NewTag(tt.args.epc)
+			require.Equal(t, got, tt.want)
+		})
+	}
+}
+
+func TestSetState(t *testing.T) {
+	type fields struct {
+		LastRead int64
+		state    TagState
+	}
+	type args struct {
+		newState TagState
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+	}{
+		{
+			name:   "OK - change to Unknown state",
+			fields: fields{state: TagState(Present), LastRead: 9999999999},
+			args:   args{newState: TagState(Unknown)},
+		},
+		{
+			name:   "OK - change to Departed state",
+			fields: fields{state: TagState(Present), LastRead: 9999999999},
+			args:   args{newState: TagState(Departed)},
+		},
+		{
+			name:   "OK - maintain Unknown state",
+			fields: fields{state: TagState(Unknown), LastRead: 9999999999},
+			args:   args{newState: TagState(Unknown)},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tag := &Tag{
+				LastRead: tt.fields.LastRead,
+				state:    tt.fields.state,
+			}
+			tag.setState(tt.args.newState)
+			require.Equal(t, tt.args.newState, tag.state)
+		})
+	}
+}
+
+func TestSetStateAt(t *testing.T) {
+	type fields struct {
+		LastDeparted int64
+		LastArrived  int64
+		state        TagState
+	}
+	type args struct {
+		newState  TagState
+		timestamp int64
+	}
+	tests := []struct {
+		name   string
+		fields fields
+		args   args
+	}{
+		{
+			name:   "OK - change to Present state",
+			fields: fields{state: Unknown, LastArrived: 9999999998, LastDeparted: 9999999999},
+			args:   args{newState: Present, timestamp: 9999999999},
+		},
+		{
+			name:   "OK - change to Departed state",
+			fields: fields{state: Unknown, LastDeparted: 9999999998, LastArrived: 9999999999},
+			args:   args{newState: Departed, timestamp: 9999999999},
+		},
+		{
+			name:   "OK - change to Unknown state",
+			fields: fields{state: Departed, LastDeparted: 9999999998, LastArrived: 9999999998},
+			args:   args{newState: Unknown, timestamp: 9999999998},
+		},
+		{
+			name:   "OK - maintain Unknown state",
+			fields: fields{state: Unknown, LastDeparted: 9999999998, LastArrived: 9999999998},
+			args:   args{newState: Unknown, timestamp: 9999999998},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tag := &Tag{
+				LastDeparted: tt.fields.LastDeparted,
+				LastArrived:  tt.fields.LastArrived,
+				state:        tt.fields.state,
+			}
+			tag.setStateAt(tt.args.newState, tt.args.timestamp)
+			require.Equal(t, tag.state, tt.args.newState)
+			require.Equal(t, tag.LastArrived, tt.args.timestamp)
+			require.Equal(t, tag.LastDeparted, tt.args.timestamp)
```

Comment:
Typically only use `require` when the rest of the tests is invalid to continue if the condition fails. In this case it is still valid to test each of these with `assert`.

Reply:
Can you elaborate on this a bit more? I'm not sure I'm following...

Reply:
Yup. The test will terminate when `require` condition fails. `assert` will fails the test if the condition fails, but not terminate the test. So we use `assert` to test conditions that are not dependent of another condition and `require` if the failing condition invalidates the rest of the test.

Reply:
We can jump on a call if need more details.

Reply:
I think I understand now, especially after reading the last comment you made on `util_test.go`. Thanks for the extra clarification! I'm making the appropriate changes now.